apps/prairielearn/elements/pl-file-upload/pl-file-upload.py

```diff
@@ -37,48 +110,89 @@ def add_format_error(data: pl.QuestionData, error_string: str) -> None:
+        )
 
     if "_required_file_names" not in data["params"]:
         data["params"]["_required_file_names"] = []
-    file_names = get_file_names_as_array(pl.get_string_attrib(element, "file-names"))
-    data["params"]["_required_file_names"].extend(file_names)
 
+    required_file_names = get_file_names_as_array(
+        pl.get_string_attrib(element, "file-names", "")
+    )
+    data["params"]["_required_file_names"].extend(required_file_names)
 
-def render(element_html: str, data: pl.QuestionData) -> str:
-    if data["panel"] != "question":
-        return ""
 
+def render(element_html: str, data: pl.QuestionData) -> str:
     element = lxml.html.fragment_fromstring(element_html)
     uuid = pl.get_uuid()
+    parse_error = data["format_errors"].get("_files", None)
+
+    if data["panel"] != "question":
+        return ""
 
     raw_file_names = pl.get_string_attrib(element, "file-names", "")
     file_names = sorted(get_file_names_as_array(raw_file_names))
     file_names_json = json.dumps(file_names, allow_nan=False)
 
-    answer_name = get_answer_name(raw_file_names)
+    raw_optional_file_names = pl.get_string_attrib(element, "optional-file-names", "")
+    optional_file_names = sorted(get_file_names_as_array(raw_optional_file_names))
+
+    # Split optional names into two separate lists: wildcard patterns and plain names
+    optional_file_regex, optional_file_plain = extract_patterns(optional_file_names)
+
+    optional_file_plain_json = json.dumps(optional_file_plain, allow_nan=False)
+    optional_file_regex_json = json.dumps(optional_file_regex, allow_nan=False)
+
+    answer_name = get_answer_name(raw_file_names, raw_optional_file_names)
 
     # Only send the file names to the client. We don't include the contents
     # to avoid bloating the HTML. The client will fetch any submitted files
-    # asynchronously once the page loads.
-    #
-    # We filter out any files that weren't specified in the file names for this element.
+    # asynchronously once the page loads
     submitted_files = data["submitted_answers"].get("_files", [])
-    submitted_file_names = list(
-        {x.get("name") for x in submitted_files} & set(file_names)
-    )
-    submitted_file_names_json = json.dumps(submitted_file_names, allow_nan=False)
+    submitted_file_names = [x.get("name") for x in submitted_files]
+
+    # We filter out any files that neither match a required file name or plain optional name
+    required_files = set(submitted_file_names) & set(file_names)
+    optional_files = set(submitted_file_names) & set(optional_file_plain)
+
+    # For wildcard patterns, we have to do a more complex pairwise matching of files and patterns
+    wildcard_files = {
+        file
+        for pattern, file in itertools.product(
+            optional_file_regex, submitted_file_names
+        )
+        if re.compile(pattern[0], re.IGNORECASE).match(file)
```

Comment:
Can you split this into a nested loop instead of a comprehension? The way this is currently written, you're doing repeated work by compiling the same pattern for each file in the list of submitted file names.

Reply:
Done.